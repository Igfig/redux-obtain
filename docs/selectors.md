# Selectors

Sometimes you have components on the other side of the React Tree that need to know about data you've fetched from the server.

## getResourceData

`getResourceData` is a simple selector that will return the `data` key from the redux store under your resources name.

Usage:

```javascript
import { getResourceData } from "redux-obtain"

export default connect(state => ({
    comments: getResourceData("COMMENTS")(state)
}))(({ comments }) => <div>{/* do with it what you will */}</div>)
```

Advanced Usage:

The second argument is a selector function that gets applied to the `data` key in your resource store.

```
Redux Store: {
    resources: {
        CURRENT_USER: {
            data: {
                firstName: 'Henry',
                lastName: 'Williamson'
            },
            loading: false,
            error: false
        }
    }
}
```

```javascript
import { getResourceData } from "redux-obtain"

export default connect(state => ({
    fullName: getResourceData("RESOURCE_NAME", data => `${data.firstName} ${data.lastName}`)(state)
}))(({ fullName }) => <span>{fullName /* Henry Williamson */}</span>)
```
