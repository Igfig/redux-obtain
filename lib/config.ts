export let config = {
    requestHeaderSelector: (state: any) => ({
        "Content-Type": "application/json",
        Authorization: `Basic ${state.authentication.token}`
    }),
    reduxStoreName: "resources"
}

export const setupFetcher = (options: any) => {
    config = {
        ...config,
        ...options
    }
}
