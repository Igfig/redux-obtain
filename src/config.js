export let config = {
    requestHeaderSelector: state => ({
        "Content-Type": "application/json",
        Authorization: `Basic ${state.authentication.token}`
    }),
    reduxStoreName: "resources",
    paginationInitialLoadLimit: 500,
    getOrderBys: props =>
        props && props.sortBy && props.sortDirection
            ? props.sortBy.map((column, index) => ({
                  column,
                  direction: props.sortDirection[index]
              }))
            : []
}

export const setupFetcher = options => {
    config = {
        ...config,
        ...options
    }
}
