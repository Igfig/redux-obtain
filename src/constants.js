export const initialState = {}
export const initialResourceState = {
    data: {},
    loading: false,
    error: false
}

const prefix = "@@fetcher"

export const ADD_RESOURCE = `${prefix}/ADD_RESOURCE`
export const MODIFY_RESOURCE = `${prefix}/MODIFY_RESOURCE`
export const REQUEST_RESOURCE = `${prefix}/REQUEST_RESOURCE`
export const FETCH_SUCCESS = `${prefix}/FETCH_SUCCESS`
export const FETCH_ADDITIONAL_SUCCESS = `${prefix}/FETCH_ADDITIONAL_SUCCESS`
export const FETCH_ERROR = `${prefix}/FETCH_ERROR`
export const REMOVE_RESOURCE = `${prefix}/REMOVE_RESOURCE`
