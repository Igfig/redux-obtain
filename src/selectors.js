import { config } from "./config"

export const getResourceData = (name, selector) => state =>
    state[config.reduxStoreName][name]
        ? selector
          ? selector(state[config.reduxStoreName][name].data)
          : state[config.reduxStoreName][name].data
        : undefined

export const getMetaData = name => state => ({
    loading: state[config.reduxStoreName][name].loading,
    error: state[config.reduxStoreName][name].error
})
