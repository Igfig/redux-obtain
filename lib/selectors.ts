import * as C from "./constants"

export const getResourceData = (name: string, selector: any) => (state: any) =>
    state[C.REDUX_STORE_NAME][name]
        ? selector
          ? selector(state[C.REDUX_STORE_NAME][name].data)
          : state[C.REDUX_STORE_NAME][name].data
        : undefined
