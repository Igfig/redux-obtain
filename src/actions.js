import * as C from "./constants"

export const addResource = (name, paginationKey) => ({
    type: C.ADD_RESOURCE,
    name,
    paginationKey
})

export const modifyResource = ({ name, dataTransform }) => ({
    type: C.MODIFY_RESOURCE,
    name,
    dataTransform
})

export const requestResource = name => ({ type: C.REQUEST_RESOURCE, name })

export const fetchSuccess = (name, data, acceptResponse) => ({
    type: C.FETCH_SUCCESS,
    name,
    data,
    acceptResponse
})

export const fetchAdditionalSuccess = (name, data) => ({
    type: C.FETCH_ADDITIONAL_SUCCESS,
    name,
    data
})

export const fetchError = (name, error) => ({ type: C.FETCH_ERROR, name, error })

export const removeResource = name => ({ type: C.REMOVE_RESOURCE, name })
