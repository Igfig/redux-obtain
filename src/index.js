import { fetcher } from "./fetcher"
import { modifyResource } from "./actions"
import { getResourceData, getMetaData } from "./selectors"
import { reducer } from "./reducer"
import { setupFetcher } from "./config"

export { fetcher, modifyResource, getResourceData, getMetaData, reducer, setupFetcher }
