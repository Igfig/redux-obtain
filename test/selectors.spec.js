import { config } from "../src/config"
import { getResourceData, getMetaData } from "../src/selectors"

describe("Selectors", () => {
    describe("getResourceData", () => {
        it("should do nothing if resource does not exist", () => {
            expect(getResourceData("some_key")({ [config.reduxStoreName]: {} })).toMatchSnapshot()
        })
        it("should select the correct data", () => {
            const data = [1, 2, 3, 4, 5]
            const RESOURCE_KEY = "LIST_OF_STUFF"
            const state = {
                [config.reduxStoreName]: {
                    [RESOURCE_KEY]: {
                        data,
                        loading: false,
                        error: false
                    }
                }
            }
            expect(getResourceData(RESOURCE_KEY)(state)).toMatchSnapshot()
        })
        it("should apply secondary selector", () => {
            const payload = "Hello world!"
            const data = { deeply: { nested: { data: payload } } }
            const RESOURCE_KEY = "whatever dude"
            const state = {
                [config.reduxStoreName]: {
                    [RESOURCE_KEY]: {
                        data,
                        loading: false,
                        error: false
                    }
                }
            }
            const selector = data => data.deeply.nested.data
            expect(getResourceData(RESOURCE_KEY, selector)(state)).toMatchSnapshot()
        })
    })
    describe("getMetaData", () => {
        it("should select the correct data", () => {
            const state = {
                [config.reduxStoreName]: {
                    rKey: {
                        data: {},
                        loading: true,
                        error: false
                    },
                    aKey: {
                        data: {},
                        loading: false,
                        error: true
                    }
                }
            }
            expect(getMetaData("rKey")(state)).toMatchSnapshot()
            expect(getMetaData("aKey")(state)).toMatchSnapshot()
        })
    })
})
