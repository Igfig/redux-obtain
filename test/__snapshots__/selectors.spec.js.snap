// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Selectors getMetaData should select the correct data 1`] = `
Object {
  "error": false,
  "loading": true,
}
`;

exports[`Selectors getMetaData should select the correct data 2`] = `
Object {
  "error": true,
  "loading": false,
}
`;

exports[`Selectors getResourceData should apply secondary selector 1`] = `"Hello world!"`;

exports[`Selectors getResourceData should do nothing if resource does not exist 1`] = `undefined`;

exports[`Selectors getResourceData should select the correct data 1`] = `
Array [
  1,
  2,
  3,
  4,
  5,
]
`;
